### DISCUSSION TITLE
#15 📑 Weekly Report 15

### DISCUSSION THREAD

**Discussion by fehlend0** [SE-TINF22B6/Statify#98] [Blog]
Hi there👋 This time we have had a shorter week, but still we were able to manage some tasks.

## Code Refactoring

This week's main task was to refactor the source code. In this [documentation](https://github.com/SE-TINF22B6/Statify/blob/main/Documentation/Code_Refactoring.md) you can get some insights of the principles we stick to. To keep our source code clean, we use:

- **Meaningful Names**: We use descriptive names for variables, functions, and classes to enhance readability and maintainability.
- **Comments**: Comments are added where code is not self-explanatory, offering insights into developers' thoughts.
- **Folder Structure**: Our project has a well-organized folder structure, aiding in managing large codebases efficiently and facilitating modular development.
- **Smaller Functions**: Large functions are decomposed into smaller ones for better understanding, testing, and debugging.
- **Named Constants**: We prefer named constants over hard-coded values for clarity and easy updates.
- **Restructuring of the Data Model**: Entities are restructured for consistency and extensibility by introducing a common base class.

In the future we will remove unused code parts as well.

## Changes in the Database Schema

We decided to improve our database structure. The database schema was changed to store the top 5 artists/tracks together as a single dataset, instead of individually. This change was made to simplify the retrieval and processing of related entries. Storing them as a set saves storage space and reduces the complexity of queries. This keeps the database more organized and makes data queries more efficient.

Here is the new data schema: 

![DataModel](https://github.com/SE-TINF22B6/Statify/assets/106541016/4f35c903-7816-49cb-b765-df0fa70a9770)

## Changes in the Frontend

In the frontend, we use Chart.js to represent some audio features of a track in a single chart. We have already implemented it with some sample data. Now, when we use real values, we see that one of the values does not fit in, since all the values vary from 0 to 1, except for loudness, which may be from -60 to 0 and would probably ruin the appearance of the chart. Therefore, we decided to exclude it.

Another enhancement we made in the frontend was the adjustment of the time limitation options that can be chosen by the user to generate statistics. Now we have three clear options: long term (1 year), medium term (6 months), and short term (4 weeks).

This is how it looks like now:

![generate](https://github.com/SE-TINF22B6/Statify/assets/106541016/d7cc8798-82b2-483c-9258-92e3afa6601a)

## Changes in the Backend

Furthermore, the task of connecting the Backend and the Frontend is almost completed. For displaying the statistics, generating new statistics as well as displaying details of a specific track, the Frontend now sends requests to the Backend which returns the requested data. Currently, the Backend returns only sample data but soon we will enhance this feature so that real data stored in our database will be returned.

Stay tuned to see, what will come next week!

Your Statify Team

**Comment by sfrancisw** [comment]
Dear statify team,

It's great how you described your refactoring steps in detail! It's also clear what you've worked on this week. Seems like you're working on something everywhere, from frontend to backend and your database. Be sure to stay motivated on the refactoring and the code cleanup!

Your team time2shine

**Comment by LenVej** [comment]
Hey Statify Team,

You've clearly laid out your principles for enhancing readability, maintainability, and extensibility.
The restructuring of your data model and the database schema update are smart moves.
Additionally, your frontend enhancements along with the ongoing backend-frontend integration show great progress. 
Keep up the excellent work!

Your CookHub Team

**Comment by HowDoIUseGit** [comment]
Hi Statify,

You've clearly laid out your principles for enhancing readability, maintainability, and extensibility. The restructuring of your data model and the database schema update are smart moves. Additionally, your frontend enhancements along with the ongoing backend-frontend integration show great progress.

Your CookPit

**Comment by Umbranecis** [comment]
Hey Stratify,

I really like the class diagram you provided, it makes it a lot easier to understand what you did.

Further more, I think despite only being a small change, the frontend change to manually choose the timeframe for stats is a huge benefit for the user. Keep up the work

Team Play
